### PR TITLE
#680 use pyproject.toml

### LINKED ISSUE #680: Modernize Packaging: pyproject.toml
Need to modernize packaging around pyproject.toml:

- https://packaging.python.org/en/latest/guides/modernize-setup-py-project/

Care to be had with backward compat.... perhaps this should be slated for 3.2. Additionally, should poetry or the like be considered?

### PR DESCRIPTION
**Pull request by sigma67**
**NOTICE:** Prior to submitting a pull request, please ensure you abide by the
[Guidelines for Code Contributions], including but not limited to:

- PR is associated with at least one issue
- Submit from a **topic** branch, not a pre-existing Cement branch
- Make every effort to comply with [PEP8]
- All tests pass successfully
- Coverage reports 100% code coverage

---

**Issue:** closes #680 

[Guidelines for Code Contributions]: https://github.com/datafolklabs/cement/blob/master/.github/CONTRIBUTING.md#guidelines-for-code-contributions
[PEP8]: http://www.python.org/dev/peps/pep-0008/

### CODE REVIEW

File: docker/Dockerfile.dev [hunk 1]
```diff
@@ -3,9 +3,8 @@ LABEL MAINTAINER="BJ Dierkes <derks@datafolklabs.com>"
 ENV PS1="\[\e[0;33m\]|> cement-py312 <| \[\e[1;35m\]\W\[\e[0m\] \[\e[0m\]# "
 
 WORKDIR /src
-COPY requirements-dev.txt /src/
-RUN apk update \
-    && apk add libmemcached-dev \
+RUN apk update --no-check-certificate \
```
Comment (sigma67) [inline docker/Dockerfile.dev:6]: The docker files are still in draft state - I just added this to make them work in my environment, will remove the no-check-certificate again

File: Makefile [hunk 2]
```diff
@@ -7,38 +7,32 @@ dev:
 	docker-compose exec cement /bin/bash
 
 test: comply
-	python -m pytest -v --cov=cement --cov-report=term --cov-report=html:coverage-report --capture=sys tests/
+	pdm run pytest --cov=cement tests
```
Comment (sigma67) [inline Makefile:10]: I moved most of the configuration for pytest to pyproject.toml

File: Dockerfile [hunk 3]
```diff
@@ -3,7 +3,7 @@ LABEL MAINTAINER="BJ Dierkes <derks@datafolklabs.com>"
 ENV PS1="\[\e[0;33m\]|> cement <| \[\e[1;35m\]\W\[\e[0m\] \[\e[0m\]# "
 WORKDIR /src
 COPY . /src
-RUN python setup.py install \
+RUN pip install . \
```
Comment (sigma67) [inline Dockerfile:6]: I think this is an integration test - but  I didn't quite understand the need to move `.` to `src`

File: .github/workflows/pythonpublish.yml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+name: Upload Python Package
```
Comment (sigma67) [inline .github/workflows/pythonpublish.yml:1]: I added this to simplify pypi uploads - I use the same workflow in my projects

All you need to do is create a release and the rest will be taken care of by pypi

All you need to do is set up trusted publishing in your pypi settings (not hard to do)

https://docs.pypi.org/trusted-publishers/

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    "E402",
+    "E713",
+    "E721",
+    "E714",
+    "F841",
+    "F507"
+]
+
+# TBD
+# [tool.ruff.format]
```
Comment (sigma67) [inline pyproject.toml:115]: considering formatting out of scope - it would introduce too many unrelated changes here

**Review by sigma67 — COMMENTED** [review]
Hi @derks , I have left some comments on the points that should be discussed

Sorry for introducing so many changes at once, but there's not really a good way to do this incrementally

File: docker/Dockerfile.dev-py38 [hunk 6]
```diff
@@ -1,9 +1,8 @@
 FROM python:3.8-alpine
 LABEL MAINTAINER="BJ Dierkes <derks@datafolklabs.com>"
-ENV PS1="\[\e[0;33m\]|> cement-py38 <| \[\e[1;35m\]\W\[\e[0m\] \[\e[0m\]# "
+ENV PS1="\[\e[0;33m\]|> cement-py8 <| \[\e[1;35m\]\W\[\e[0m\] \[\e[0m\]# "
```
Comment (sigma67) [inline docker/Dockerfile.dev-py38:3]:
```suggestion
ENV PS1="\[\e[0;33m\]|> cement-py38 <| \[\e[1;35m\]\W\[\e[0m\] \[\e[0m\]# "
```

**Comment by derks** [comment]
@sigma67 really appreciate your time on this. As always, apologies for the delays in review... but plan to get into this as soon as I can.

**Comment by sigma67** [comment]
Thanks for the adjustments & merge!
